### PR TITLE
Import automation: import attempt resource

### PR DESCRIPTION
**Pull request by intrepiditee**
An import attempt resource is conceptually a dict that has keys like import_name and logs. It describes the progress of an import attempt and is uniquely identified by its attempt_id.

Endpoint '/import/<string:attempt_id>' is exposed for managing import attempts and '/imports' is exposed for querying import attempts in bulk.

### CODE REVIEW

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        """
+        import_attempt = self.database.get_by_id(attempt_id)
+        if not import_attempt:
+            return NOT_FOUND_ERROR, http.HTTPStatus.NOT_FOUND
+        return import_attempt
+
+    def patch(self, attempt_id):
+        """Modifies the value of a field of an existing attempt.
+
+        attempt_id is forbidden to modify.
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py:120]: I  do not understand this sentence.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+        Args:
+            attempt_id: ID of the attempt as a string
+
+        Returns:
+            The import attempt with the ID if successful as a dict. Otherwise,
+            (error message, error code), where the error message is a string
+            and the error code is an int.
+        """
+        args = self.parser.parse_args()
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py:130]: What values does "args" take?  I do not understand the function call.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+
+class ImportAttemptList(ImportAttempt):
+    """API for querying a list of import attempts based on some criteria."""
+
+    def get(self):
+        """Retrieves a list of import attempts that pass the filter defined by
+        the key-value mappings in the request body."""
+        args = self.parser.parse_args()
+        imports = self.database.filter(args)
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py:148]: Just return this directly.

File: import-automation/progress-dashboard-rest/test/test_import_attempt.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Tests for import_attempt.py.
+"""
+
+
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/test/test_import_attempt.py:19]: Likewise on reducing vertical whitespace.

File: import-automation/progress-dashboard-rest/test/test_import_attempt.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+import unittest
+from unittest import mock
+
+
+PARSE_ARGS = 'flask_restful.reqparse.RequestParser.parse_args'
+IMPORT_ATTEMPT_DATABASE = 'app.resource.import_attempt' \
+                          '.import_attempt_database.ImportAttemptDatabase'
+EXAMPLE_ATTEMPT = {
+    'attempt_id': '0', 'branch_name': 'branch',
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/test/test_import_attempt.py:32]: Please have one entry per line, easing human comprehension.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Import attempt resources associated with the endpoints
+'/import/<string:attempt_id>' and '/imports'.
+"""
+
+
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:20]: fixed

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+from enum import Enum
+
+import flask_restful
+from flask_restful import reqparse
+
+from ..service import import_attempt_database
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:26]: fixed

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+from .. import utils
+
+import http
+
+
+class ImportStatus(Enum):
+    """Allowed status of an import attempt."""
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:33]: The status of an import attempt can only be one of these. Added this sentence

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    CREATED = 'created'
+    SUCCEEDED = 'succeeded'
+    FAILED = 'failed'
+
+
+IMPORT_STATUS = set(status.value for status in ImportStatus)
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:39]: None. The user of the API can change the status of an attempt at will.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+ID_NOT_MATCH_ERROR = 'attempt_id in path variable does not match attempt_id ' \
+                     'in request body '
+NOT_FOUND_ERROR = 'attempt_id not found'
+
+
+class ImportAttempt(flask_restful.Resource):
+    """Base import attempt resource.
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:47]: Base class of an import attempt resource

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 1 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:120]: It is not allowed to modify the value of attempt_id

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 2 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:130]: The parser is inherited from the base class ImportAttempt. The parser looks for fields defined in the base class.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 3 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:148]: fixed

File: import-automation/progress-dashboard-rest/test/test_import_attempt.py (hunk 4 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/test/test_import_attempt.py:19]: fixed

File: import-automation/progress-dashboard-rest/test/test_import_attempt.py (hunk 5 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/test/test_import_attempt.py:32]: fixed

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 10 above)
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py]:
```suggestion
    """Base class for an import attempt resource.
```

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 1 above)
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py]: Perhaps change the argument specification to "as a constant string", omitting this comment.

File: import-automation/progress-dashboard-rest/test/test_import_attempt.py (hunk 5 above)
Comment (jeffreyoldham) [reply import-automation/progress-dashboard-rest/test/test_import_attempt.py]: I do not see the changes, but perhaps they are elsewhere.  I am still learning the GitHub UI.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 11]
```diff
@@ -117,17 +121,17 @@ def get(self, attempt_id):
     def patch(self, attempt_id):
         """Modifies the value of a field of an existing attempt.
 
-        attempt_id is forbidden to modify.
+        It is not allowed to modify the value of attempt_id.
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py:124]: Perhaps drop this comment in favor of "as a constant string" below.

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py [hunk 12]
```diff
@@ -117,17 +121,17 @@ def get(self, attempt_id):
...
 
         Args:
-            attempt_id: ID of the attempt as a string
+            attempt_id: ID of theI attempt as a string
```
Comment (jeffreyoldham) [inline import-automation/progress-dashboard-rest/app/resource/import_attempt.py:127]:
```suggestion
            attempt_id: ID of the attempt as a constant string.
```

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 11 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:124]: I was trying to that the ID of an import attempt could not be changed not the input

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 10 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py]: fixed

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 11 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py:124]: Maybe "The attempt_id of an existing import attempt resource is forbidden to be patched."?

File: import-automation/progress-dashboard-rest/app/resource/import_attempt.py (hunk 12 above)
Comment (intrepiditee) [reply import-automation/progress-dashboard-rest/app/resource/import_attempt.py]: fixed theI

**Pull request merged**